kernel/preemption/src/lib.rs

```diff
@@ -0,0 +1,115 @@
+//! Manages preemption on a per-CPU core basis.
+//! 
+//! Supports enabling and disabling preemption for the purpose of 
+//! safe task state management, e.g., through preemption-safe locks.
+
+#![no_std]
+
+extern crate alloc;
+
+use core::sync::atomic::{AtomicU8, Ordering};
+
+/// Theseus uses a `u8` to hold each CPU core's ID (apic ID),
+/// so the maximum number of cores it supports is `u8::MAX` (256).
+const MAX_CPU_CORES: usize = u8::MAX as usize;
+
+const ATOMIC_U8_ZERO: AtomicU8 = AtomicU8::new(0);
+
+/// The per-core preemption count, indexed by a CPU core's APIC ID.
+/// 
+/// If a CPU's count is `0`, preemption is enabled.
+/// If a CPU's count is greater than `0`, preemption is disabled.
+static PREEMPTION_COUNT: [AtomicU8; MAX_CPU_CORES] = [ATOMIC_U8_ZERO; MAX_CPU_CORES];
+
+
+/// Prevents preemption (preemptive task switching) from occurring
+/// until the returned guard object is dropped.
+pub fn hold_preemption() -> PreemptionGuard {
+    let apic_id = apic::get_my_apic_id();
+    let prev_val = PREEMPTION_COUNT[apic_id as usize].fetch_add(1, Ordering::Relaxed);
+    // If the previous counter value was 0, that indicates we are transitioning
+    // from preemption being enabled to disabled on this CPU.
+    let preemption_was_enabled = prev_val == 0;
+    // Create a guard here immediately after incrementing the counter,
+    // in order to guarantee that a failure below will drop it and decrement the counter.
+    let guard = PreemptionGuard {
+        apic_id,
+        preemption_was_enabled,
+    };
+
+    if preemption_was_enabled {
+        // When transitioning from preemption being enabled to disabled,
+        // we must disable the local APIC timer used for preemptive task switching.
+        apic::get_my_apic()
+            .expect("BUG: hold_preemption() couldn't get local APIC")
+            .write()
+            .enable_lvt_timer(false);
+    } else if prev_val == u8::MAX {
+        // Overflow occurred and the counter value wrapped around, which is a bug.
+        panic!("BUG: Overflow occurred in the preemption counter for CPU {}", apic_id);
+    }
+
+    guard
+}
+
+
+/// A guard type that ensures preemption is disabled as long as it is held.
+/// 
+/// Call [`hold_preemption()`] to obtain a `PreemptionGuard`.
+/// 
+/// Preemption *may* be re-enabled when this guard is dropped,
+/// but not necessarily so, because other tasks on this CPU 
+/// may also have acquired a `PreemptionGuard`.
+pub struct PreemptionGuard {
+    /// The APIC ID of the CPU on which preemption was held.
+    /// This is not necessary, it's just used for sanity checks or debugging.
+    apic_id: u8,
```

Comment:
Should we only enable this field (and any modifications of this field) in debug builds?
```suggestion
    /// The APIC ID of the CPU on which preemption was held.
    ///
    /// This is not necessary; it's just used for sanity checks or debugging.
    #[cfg(debug_assertions)
    apic_id: u8,
```

Reply:
i would, but we basically never build in debug mode because it runs prohibitively slowly. Considering that we're using it for sanity checks now, it's best to just change the comment to reflect that.